### PR TITLE
Add personal notify feature

### PR DESCRIPTION
**Pull request by volo-zyko**
With this code `ZulipNotifier` along with messages to some stream/topic will send personal notifications to code committers.

I didn't update `README.markdown` yet but I surely do it once you decide that the changes are worthy.

Another thing. `zulipPersonalSend` similar to `zulipSend` can be useful and I can add it too.

### CODE REVIEW

**Comment by volo-zyko** [comment]
CI has failed with some unclear messages which, it seems, are not related to my changes.

Windows:
`Remote call on JNLP4-connect connection from 52.191.116.151/52.191.116.151:49187 failed`

Linux:
`java.lang.OutOfMemoryError: Java heap space`

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java [hunk 1]
```diff
@@ -74,73 +83,170 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
...
         StringBuilder changeString = new StringBuilder();
+        if (changes == null) {
+            changeString.append("Could not determine changes since the last build.");
+        } else if (!changes.isEmpty()) {
+            // If there seems to be a commit message at all, try to list all the changes.
+            changeString.append("Changes since the last build:\n");
+            for (CommitInfo change : changes) {
+                if (ZulipUtil.isValueSet(change.email)) {
+                    changeString.append("\n* ");
+                    changeString.append(hundsonUrlMesssage("", change.author, "mailto:" + change.email));
```
Comment (volo-zyko) [inline src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: better ```"`" + change.author + "`"``` to make the formatting similar to to the one when email is not set.

**Comment by butchyyyy** [comment]
Hi @volo-zyko , thank you for the contribution! I'm on my vacation now so I only took a brief look. I'll do a thorough review next week. In general I like the possibility to nag people who broke the build via PM :)

The failing build is probably just something in the CI infrastructure so don't worry about it now. I'll run the build / tests as a part of the review.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java [hunk 2]
```diff
@@ -74,73 +83,170 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
         publish(run, listener);
     }
 
+    static private class CommitInfo {
```
Comment (butchyyyy) [inline src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: I find having this as a static private class confusing. Can you make this a regular class (pojo)?

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java [hunk 3]
```diff
@@ -74,73 +83,170 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
...
+                    }
+                    zulip.sendPrivateMessage(personalChangeString.getKey(), privateMessage);
+                }
+            }
         }
         return true;
     }
 
-    private String getChangeSet(Run<?, ?> build) {
+    private Map<String, StringBuilder> buildPersonalChangeStrings(List<CommitInfo> changes) {
```
Comment (butchyyyy) [inline src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: I think the committer should see the **full** change set instead of just his commits, since any of those could have broken the build.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java [hunk 4]
```diff
@@ -36,6 +42,9 @@
     @Extension
     public static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
 
+    public static final Pattern EMAIL_PATTERN =
```
Comment (butchyyyy) [inline src/main/java/jenkins/plugins/zulip/ZulipNotifier.java:54]: The pattern does not cover all possible characters that can be in both local part and domain part of the emails. While this will work in most cases, maybe there's an utility for that out there?

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java [hunk 5]
```diff
@@ -74,73 +83,170 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
...
-            // If there seems to be a commit message at all, try to list all the changes.
-            changeString.append("Changes since last build:\n");
             for (ChangeLogSet<? extends ChangeLogSet.Entry> changeLogSet : wrapper.getChangeSets()) {
                 for (ChangeLogSet.Entry e : changeLogSet) {
                     String commitMsg = e.getMsg().trim();
                     if (commitMsg.length() > 47) {
                         commitMsg = commitMsg.substring(0, 46) + "...";
                     }
+                    // GitChangeSet uses email as author's Id.
+                    String email = "";
```
Comment (butchyyyy) [inline src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: Did you manage to test this on both Git and Mercurial?

Can you create an utility functions that will extract email / display name from the commit info? The current implementation is hard to read. It will also be easier to write thorough tests for the detection.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java [hunk 6]
```diff
@@ -74,73 +83,170 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
...
         StringBuilder changeString = new StringBuilder();
+        if (changes == null) {
+            changeString.append("Could not determine changes since the last build.");
+        } else if (!changes.isEmpty()) {
+            // If there seems to be a commit message at all, try to list all the changes.
+            changeString.append("Changes since the last build:\n");
+            for (CommitInfo change : changes) {
+                if (ZulipUtil.isValueSet(change.email)) {
+                    changeString.append("\n* ");
+                    changeString.append(hundsonUrlMesssage("", "`" + change.author + "`", "mailto:" + change.email));
```
Comment (butchyyyy) [inline src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: Since this is a Zulip integration, would't it be better to be able to message the person via Zulip instead of sending him an e-mail?

While related, this is not part of the personal notify feature and should be in a separate commit. Please rebase to keep the history clean.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java [hunk 7]
```diff
@@ -174,16 +280,16 @@ private static boolean shouldPublish(Run<?, ?> build) {
      * @param prefix       The prefix to add to the parameter (will be formatted as bold)
      * @param display      The build parameter
      * @param url          The Url to the Jenkins item
-     * @param globalConfig Global step config
      * @return Formatted parameter
      */
-    private static String hundsonUrlMesssage(String prefix, String display, String url, DescriptorImpl globalConfig) {
+    private static String hundsonUrlMesssage(String prefix, String display, String url) {
```
Comment (butchyyyy) [inline src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: Changes that are not part of the smart notification feature should be in a separate commit (or PR).

Never the less, i don't think you should hijack function that was designed to link to Jenkins project / build pages to generate mailto link. I think this should be reverted.

File: src/main/webapp/help-globalConfig-personalNotify.html [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>When checked, build notifications will be posted to committers as private messages.</p>
```
Comment (butchyyyy) [inline src/main/webapp/help-globalConfig-personalNotify.html:2]: Since the whole personal notification idea is based around committer email, i'm guessing this won't work for users that logged in to Zulip via Github or Google?
If that's the case we should definitely let user know in the help / docs / readme.

**Review by butchyyyy — CHANGES_REQUESTED** [review]
Sorry it took so long to review this. Was super busy lately. I've posted few questions / remarks.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java (hunk 2 above)
Comment (volo-zyko) [reply src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: I'm not a Java programmer, so excuse my stupid questions. Do you mean to move this class outside ZulipNotifier class?

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java (hunk 4 above)
Comment (volo-zyko) [reply src/main/java/jenkins/plugins/zulip/ZulipNotifier.java:54]: I'll try to find one. This is a common problem with emails. The format defined by RFC is overly complex but at the same time there is no standard parsing facility for it. So, people normally come up with something which almost works. :)

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java (hunk 6 above)
Comment (volo-zyko) [reply src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: Yes, I was thinking about it but I couldn't find how to mention a user knowing his/her email. Even when Zulip id is the same as id on Jenkins I don't know how to mention that user in a message.

File: src/main/webapp/help-globalConfig-personalNotify.html (hunk 8 above)
Comment (volo-zyko) [reply src/main/webapp/help-globalConfig-personalNotify.html:2]: That's true. I'll document that.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java (hunk 7 above)
Comment (volo-zyko) [reply src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: In general I didn't break this code. I extended it a bit but the result was the same. So, links to the project / build pages were still generated.

However, I decided to revert this code. My main reason for changing it was overly long lines with the committer info. If author has a long name/surname then in some setups it will be duplicated in the messages sent to the Zulip server. Consider this message:

```
* Johnathan Kristoffersen <johnathan.kristoffersen@example.com> Add test
```

So, I thought that `johnathan.kristoffersen@example.com` can be hidden behind an URL.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java (hunk 2 above)
Comment (butchyyyy) [reply src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: Yes, exactly that and good job considering you're not a Java developer.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java (hunk 4 above)
Comment (butchyyyy) [reply src/main/java/jenkins/plugins/zulip/ZulipNotifier.java:54]: Any luck with this? Should I take a look?

Comment (volo-zyko) [reply src/main/java/jenkins/plugins/zulip/ZulipNotifier.java:54]: Sorry, for the delay. I fixed all review comments within a week but then didn't have time to test things manually on a local Jenkins instance. I'll push the new code shortly.

File: src/main/java/jenkins/plugins/zulip/ZulipNotifier.java (hunk 6 above)
Comment (volo-zyko) [reply src/main/java/jenkins/plugins/zulip/ZulipNotifier.java]: Now, this can be resolved as I removed non-related changes from the PR.

**Comment by volo-zyko** [comment]
@butchyyyy, I don't understand what's wrong with my change and why it's not accepted. Is the idea itself totally wrong, or there are problems with the implementation. I fixed all the issues more than a year ago but there is no feedback. I don't understand how this works.

**Comment by volo-zyko** [comment]
It looks like the maintainer is not interested in this feature. So, I'm dropping it.
